--- ltconfig.orig	Fri Jan 19 17:48:20 2001
+++ ltconfig	Fri Jan 19 17:49:07 2001
@@ -846,10 +846,7 @@
 fi
 CFLAGS="$save_CFLAGS"
 chmod u+w .
-$rm conftest* out/*
-rmdir out
 cd ..
-rmdir conftest
 $rm -r conftest 2>/dev/null
 
 if test x"$compiler_c_o" = x"yes"; then
@@ -1333,7 +1330,7 @@
     old_archive_from_new_cmds='true'
     # FIXME: Should let the user specify the lib program.
     old_archive_cmds='lib /OUT:$oldlib$oldobjs'
-    fix_srcfile_path='`cygpath -w $srcfile`'
+    fix_srcfile_path='`CYGWIN= cygpath -w "$srcfile"`'
     ;;
 
   freebsd1*)
