Index: ltmain.sh
===================================================================
RCS file: /export/cvsroot/Coin/cfg/ltmain.sh,v
retrieving revision 1.1
diff -u -r1.1 ltmain.sh
--- ltmain.sh	2001/01/10 15:06:30	1.1
+++ ltmain.sh	2001/01/10 16:18:03
@@ -1698,15 +1698,16 @@
 	irix)
 	  major=`expr $current - $age + 1`
 	  versuffix="$major.$revision"
-	  verstring="sgi$major.$revision"
+	  verstring="sgi$major"
 
-	  # Add in all the interfaces that we are compatible with.
-	  loop=$revision
-	  while test $loop != 0; do
-	    iface=`expr $revision - $loop`
-	    loop=`expr $loop - 1`
-	    verstring="sgi$major.$iface:$verstring"
-	  done
+#	NB: This part is commented out to fix a problem with IRIX ld.
+#	  # Add in all the interfaces that we are compatible with.
+#	  loop=$revision
+#	  while test $loop != 0; do
+#	    iface=`expr $revision - $loop`
+#	    loop=`expr $loop - 1`
+#	    verstring="sgi$major.$iface:$verstring"
+#	  done
 	  ;;
 
 	linux)
